### DISCUSSION TITLE
apiservice status true, but /apis/external.metrics.k8s.io/v1beta1/ not found

### DISCUSSION THREAD

**Discussion by gigi-at-zymergen** [kedacore/keda#1698] [Q&A / Need Help]
We use Flux to apply Kubernetes manifests including the Keda manifests. Once we started using Flux to start applying Keda2 manifest, we're getting sync errors. We are observing an issue similar to https://github.com/fluxcd/flux/issues/1951, where Flux is logging the following error :

```
ts=2021-03-25T03:21:02.663577928Z caller=main.go:276 type="internal kubernetes error" kubernetes_caller=k8s.io/client-go@v0.17.4/discovery/cached/memory/memcache.go:199 err="couldn't get resource list for external.metrics.k8s.io/v1beta1: Got empty response for: external.metrics.k8s.io/v1beta1"
```

I've checked the apiService status and it looks OK :
```
$ kubectl get apiservice v1beta1.external.metrics.k8s.io -o yaml
apiVersion: apiregistration.k8s.io/v1
kind: APIService
metadata:
  annotations:
    fluxcd.io/sync-checksum: e97083f536ffb5c8f9b096037e7cde536936f128
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"apiregistration.k8s.io/v1","kind":"APIService","metadata":{"annotations":{"fluxcd.io/sync-checksum":"e97083f536ffb5c8f9b096037e7cde536936f128"},"labels":{"app.kubernetes.io/name":"v1beta1.external.metrics.k8s.io","app.kubernetes.io/part-of":"keda-operator","app.kubernetes.io/version":"2.2.0","environment":"zre_infra","fluxcd.io/sync-gc-mark":"sha256.QdyI5mN2VE3Gd1Xfdog2xdpfkEPERSAK6KUqCyKqwYI","owner":"zre"},"name":"v1beta1.external.metrics.k8s.io"},"spec":{"group":"external.metrics.k8s.io","groupPriorityMinimum":100,"insecureSkipTLSVerify":true,"service":{"name":"keda-metrics-apiserver","namespace":"tools"},"version":"v1beta1","versionPriority":100}}
  creationTimestamp: "2021-03-24T18:59:28Z"
  labels:
    app.kubernetes.io/name: v1beta1.external.metrics.k8s.io
    app.kubernetes.io/part-of: keda-operator
    app.kubernetes.io/version: 2.2.0
    environment: zre_infra
    fluxcd.io/sync-gc-mark: sha256.QdyI5mN2VE3Gd1Xfdog2xdpfkEPERSAK6KUqCyKqwYI
    owner: zre
  name: v1beta1.external.metrics.k8s.io
  resourceVersion: "171735116"
  selfLink: /apis/apiregistration.k8s.io/v1/apiservices/v1beta1.external.metrics.k8s.io
  uid: 3cf4d8e0-2bf1-4d1d-9cd3-1bc175462712
spec:
  group: external.metrics.k8s.io
  groupPriorityMinimum: 100
  insecureSkipTLSVerify: true
  service:
    name: keda-metrics-apiserver
    namespace: tools
    port: 443
  version: v1beta1
  versionPriority: 100
status:
  conditions:
  - lastTransitionTime: "2021-03-24T18:59:40Z"
    message: all checks passed
    reason: Passed
    status: "True"
    type: Available
```

When I try to get built in metrics api I am able to:
```
$kubectl get --raw "/apis/metrics.k8s.io/v1beta1/"
{"kind":"APIResourceList","apiVersion":"v1","groupVersion":"metrics.k8s.io/v1beta1","resources":[{"name":"nodes","singularName":"","namespaced":false,"kind":"NodeMetrics","verbs":["get","list"]},{"name":"pods","singularName":"","namespaced":true,"kind":"PodMetrics","verbs":["get","list"]}]}
```

However, trying to get the `external.metrics.k8s.io` api I get a resource not found
```
$  kubectl get --raw "/apis/external.metrics.k8s.io/v1beta1/" -n tools
Error from server (NotFound): the server could not find the requested resource
```
Is there something I need to do to fully register the external metrics api?

**Comment by gigi-at-zymergen** [comment]
After discussing with my team, we realized that we use rancher to manage our K8s clusters and it is proxying the kubectl requests, so that is the issue with `--raw`, we're getting the response from our Rancher cluster, not the cluster where Keda is running.

**Comment by gigi-at-zymergen** [comment]
There is an issue that if no external.metrics are defined, the call against the API does return an empty resources list, which causes errors with the [Go Kubernetes client ](https://github.com/kubernetes/client-go/blob/release-11.0/discovery/cached/memory/memcache.go#L215). Errors disappeared once we registered a dummy prometheus scaledObject.

**Reply by zroubalik** [reply]
> There is an issue that if no external.metrics are defined, the call against the API does return an empty resources list, which causes errors with the [Go Kubernetes client ](https://github.com/kubernetes/client-go/blob/release-11.0/discovery/cached/memory/memcache.go#L215). Errors disappeared once we registered a dummy prometheus scaledObject.

@gigi-at-zymergen interesting, good catch. Could you please open an issue for this? We should modify our Metrics Adapter to fix this.

**Comment by stalin4suse** [comment]
@zroubalik and @gigi-at-zymergen. Do we have link to the pull request?
I do not see this issue discussed under issues or PR.
